clang/lib/CodeGen/MicrosoftCXXABI.cpp

```diff
@@ -1144,11 +1167,13 @@ bool MicrosoftCXXABI::classifyReturnType(CGFunctionInfo &FI) const {
   if (!RD)
     return false;
 
-  bool isTrivialForABI = RD->canPassInRegisters() &&
-                         isTrivialForMSVC(RD, FI.getReturnType(), CGM);
-
   // MSVC always returns structs indirectly from C++ instance methods.
-  bool isIndirectReturn = !isTrivialForABI || FI.isInstanceMethod();
+  bool isIndirectReturn = FI.isInstanceMethod();
```

Comment:
I intentionally moved this around to try to reduce calling `isTrivialForMSVC` for methods and potentially now hitting the recursive field loop I added above.
Since methods always return indirectly we can avoid all these extra checks up front.

Let me know if you prefer this being a separate PR.

Reply:
I'd like to keep ABI changes as simple as possible; please separate.